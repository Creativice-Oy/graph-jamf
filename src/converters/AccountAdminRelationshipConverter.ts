import {
  ACCOUNT_ADMIN_RELATIONSHIP_CLASS,
  ACCOUNT_ADMIN_RELATIONSHIP_TYPE,
  ACCOUNT_ENTITY_TYPE,
  AccountAdminRelationship,
  ADMIN_ENTITY_TYPE,
} from "../jupiterone/entities";
import { Account, Admin } from "../types";
import { generateEntityKey, generateRelationKey } from "../utils/generateKey";

export function createAccountAdminRelationships(
  account: Account,
  admins: Admin[],
): AccountAdminRelationship[] {
  const defaultValue: AccountAdminRelationship[] = [];

  return admins.reduce((acc, admin) => {
    const parentKey = generateEntityKey(ACCOUNT_ENTITY_TYPE, account.id);
    const childKey = generateEntityKey(ADMIN_ENTITY_TYPE, admin.id);
    const relationKey = generateRelationKey(
      parentKey,
      ACCOUNT_ADMIN_RELATIONSHIP_TYPE,
      childKey,
    );

    const relationship: AccountAdminRelationship = {
      _class: ACCOUNT_ADMIN_RELATIONSHIP_CLASS,
      _fromEntityKey: parentKey,
      _key: relationKey,
      _type: ACCOUNT_ADMIN_RELATIONSHIP_TYPE,
      _toEntityKey: childKey,
    };

    return [...acc, relationship];
  }, defaultValue);
}
