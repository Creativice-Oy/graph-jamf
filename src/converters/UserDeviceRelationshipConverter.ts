import { User } from "../jamf";
import {
  MOBILE_DEVICE_ENTITY_TYPE,
  USER_DEVICE_RELATIONSHIP_CLASS,
  USER_DEVICE_RELATIONSHIP_TYPE,
  USER_ENTITY_TYPE,
  UserDeviceRelationship,
} from "../jupiterone/entities";
import { generateEntityKey, generateRelationKey } from "../utils/generateKey";

export function createUserDeviceRelationships(
  users: User[],
): UserDeviceRelationship[] {
  const defaultValue: UserDeviceRelationship[] = [];

  return users.reduce((acc, user) => {
    if (isMobileDevice(user)) {
      return acc;
    }

    const parentKey = generateEntityKey(USER_ENTITY_TYPE, user.id);
    const childKey = generateEntityKey(
      MOBILE_DEVICE_ENTITY_TYPE,
      user.links!.mobile_devices!.mobile_device!.id,
    );
    const relationKey = generateRelationKey(
      parentKey,
      USER_DEVICE_RELATIONSHIP_CLASS,
      childKey,
    );

    const relationship: UserDeviceRelationship = {
      _class: USER_DEVICE_RELATIONSHIP_CLASS,
      _type: USER_DEVICE_RELATIONSHIP_TYPE,
      _fromEntityKey: parentKey,
      _key: relationKey,
      _toEntityKey: childKey,
    };

    return [...acc, relationship];
  }, defaultValue);
}

function isMobileDevice(user: User) {
  return !(
    user.links &&
    user.links.mobile_devices &&
    user.links.mobile_devices.mobile_device
  );
}
