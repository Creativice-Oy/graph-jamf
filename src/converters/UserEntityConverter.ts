import { User } from "../jamf/JamfClient";
import { USER_ENTITY_CLASS, USER_ENTITY_TYPE, UserEntity } from "../jupiterone";

import generateKey from "../utils/generateKey";

export function createUserEntities(data: User[]): UserEntity[] {
  return data.map(user => {
    return {
      _key: generateKey(USER_ENTITY_TYPE, user.id),
      _type: USER_ENTITY_TYPE,
      _class: USER_ENTITY_CLASS,
      id: user.id,
      username: user.name,
    };
  });
}
