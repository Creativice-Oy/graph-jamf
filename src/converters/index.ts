export { createUserEntities } from "./UserEntityConverter";
export { createMobileDeviceEntities } from "./MobileDeviceEntityConverter";
export {
  createUserDeviceRelationships,
} from "./UserDeviceRelationshipConverter";
export { createAccountEntity } from "./AccountEntityConverter";
export {
  createAccountUserRelationships,
} from "./AccountUserRelationshipConverter";
export { createComputerEntities } from "./ComputerEntityConverter";
export {
  createUserComputerRelationships,
} from "./UserComputerRelationshipConverter";
