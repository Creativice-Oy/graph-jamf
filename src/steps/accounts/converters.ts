import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Account } from '../../jamf/types';
import { generateEntityKey } from '../../util/generateKey';
import { Entities } from '../constants';

export function getAccountEntityKey(accountId: string) {
  return generateEntityKey(Entities.ACCOUNT._type, accountId);
}

export function createAccountEntity(data: Account) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.ACCOUNT._class,
        _type: Entities.ACCOUNT._type,
        _key: getAccountEntityKey(data.id),
        displayName: data.name,
        name: data.name,
      },
    },
  });
}
