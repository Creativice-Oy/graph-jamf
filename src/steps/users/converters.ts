import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Admin, User } from '../../jamf/types';
import { generateEntityKey } from '../../util/generateKey';
import { skippedRawDataSource } from '../../util/graphObject';
import { Entities } from '../constants';

export function createAdminEntity(data: Admin) {
  return createIntegrationEntity({
    entityData: {
      source: skippedRawDataSource,
      assign: {
        _class: Entities.USER_ADMIN._class,
        _type: Entities.USER_ADMIN._type,
        _key: generateEntityKey(Entities.USER_ADMIN._type, data.id),
        id: data.id.toString(),
        username: data.name,
        displayName: data.name,
        name: data.name,
        admin: true,
        directoryUser: data.directory_user,
        fullName: data.full_name,
        email: data.email?.toLowerCase() || undefined,
        emailAddress: data.email_address?.toLowerCase(),
        active: data.enabled === 'Enabled',
        enabled: data.enabled,
        forcePasswordChange: data.force_password_change,
        accessLevel: data.access_level,
        privilegeSet: data.privilege_set,
        permissions:
          data.privilege_set === 'Administrator'
            ? [data.privilege_set]
            : undefined,
      },
    },
  });
}

export function getUserOperatingSystems(data: User) {
  const os: string[] = [];

  if (data.links) {
    if (data.links.mobile_devices.length > 0) {
      os.push('iOS');
    }

    if (data.links.computers.length > 0) {
      os.push('macOS');
    }
  }

  return os;
}

export function createDeviceUserEntity(data: User) {
  return createIntegrationEntity({
    entityData: {
      source: skippedRawDataSource,
      assign: {
        _class: Entities.DEVICE_USER._class,
        _type: Entities.DEVICE_USER._type,
        _key: generateEntityKey(Entities.DEVICE_USER._type, data.id),
        id: data.id.toString(),
        displayName: data.name,
        name: data.name,
        username: data.name,
        fullName: data.full_name,
        // Jamf device users are always considered active. Jamf _admins_ can
        // be explicitly disabled
        active: true,
        email: data.email?.toLowerCase() || undefined,
        emailAddress: data.email_address?.toLowerCase(),
        phoneNumber: data.phone_number,
        position: data.position,
        enableCustomPhotoUrl: data.enable_custom_photo_url,
        customPhotoUrl: data.custom_photo_url,
        ldapServer: data.ldap_server && data.ldap_server.name,
        totalVppCodeCount: data.links && data.links.total_vpp_code_count,
        os: getUserOperatingSystems(data),
      },
    },
  });
}
