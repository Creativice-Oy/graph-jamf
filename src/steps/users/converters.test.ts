import { createAdminEntity, createDeviceUserEntity } from './converters';
import { createMockAdminUser, createMockDeviceUser } from '../../../test/mocks';

describe('#createAdminEntity', () => {
  test('should convert data', () => {
    expect(createAdminEntity(createMockAdminUser())).toMatchSnapshot();
  });
});

describe('#createDeviceUserEntity', () => {
  test('should convert data', () => {
    expect(createDeviceUserEntity(createMockDeviceUser())).toMatchSnapshot();
  });
});
