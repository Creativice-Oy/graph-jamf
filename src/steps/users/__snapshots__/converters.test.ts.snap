// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createAdminEntity should convert data 1`] = `
Object {
  "_class": Array [
    "User",
  ],
  "_key": "jamf_user_1",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "uploadStatus": "SKIPPED",
        "uploadStatusReason": "Raw data currently disabled for this entity type",
      },
    },
  ],
  "_type": "jamf_user",
  "accessLevel": "Full Access",
  "admin": true,
  "createdOn": undefined,
  "directoryUser": false,
  "displayName": "adam.williams",
  "email": "adam.williams@jupiterone.com",
  "emailAddress": "adam.williams@jupiterone.com",
  "enabled": "Enabled",
  "forcePasswordChange": false,
  "fullName": "",
  "id": "1",
  "name": "adam.williams",
  "permissions": Array [
    "Administrator",
  ],
  "privilegeSet": "Administrator",
  "username": "adam.williams",
}
`;

exports[`#createDeviceUserEntity should convert data 1`] = `
Object {
  "_class": Array [
    "User",
  ],
  "_key": "device_user_1",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "uploadStatus": "SKIPPED",
        "uploadStatusReason": "Raw data currently disabled for this entity type",
      },
    },
  ],
  "_type": "device_user",
  "createdOn": undefined,
  "customPhotoUrl": undefined,
  "displayName": "Test",
  "email": "test@gmail.com",
  "emailAddress": "test@gmail.com",
  "enableCustomPhotoUrl": undefined,
  "fullName": "Test User",
  "id": "1",
  "ldapServer": "None",
  "name": "Test",
  "os": Array [],
  "phoneNumber": "",
  "position": "",
  "totalVppCodeCount": 0,
  "username": "Test",
}
`;
